### ISSUE TITLE
Expose RNG (again)

### ISSUE THREAD

**Issue by ablaom** [Evovest/EvoTrees.jl#52]
I see that the new MLJ models only expose a *seed* rather than the RNG. Is there a reason for this restriction?

To generate multiple learning curves for an MLJ model, one needs access to the RNG.

For convenience, setting the RNG field to an integer `i` could instantiate a `MersenneTwister(i)`. (In MLJ, most `rng` fields or key-words can be integers or `AbstractRNG`)

**Comment by jeremiedb** [comment]
As for EvoTrees, the approach I've taken has been to provide with a `seed` integer to control the reproducibility and handle multi learning curves. This borrows from what is found in XGBoost and LightGBM, so it wasn't meant as a restriction but to match what appears like the general convention for controlling the stochasticity. 

I'm not sure to get what is the limitation from using a `seed` provided as a hyper-parameter that is then used to control the stochasticity by setting the seed with `Random.seed!(params.seed)` as it does provides with the ability to generate multiple learning curves. Is the idea to provide the option for other random generators than the default one? At least in the context of GBT, I fail to see scenarios where such option would be needed. Maybe I'm missing an implication of the `seed!()` behavior?

**Comment by ablaom** [comment]
> Is the idea to provide the option for other random generators than the default one?

Yes. There I think there is a problem with reproducibility using `seed!` in multithreading, for example. If I run EvoTree tasks on different threads asynchronously, the behaviour is potentially different every time because the tasks are sharing the one global RNG.
Another example is a recent change in the behaviour of the global RNG in julia 1.4. This means different julia versions behave differently, which is an issue in testing packages that are supposed to work for multiple versions. The work-around for that is to never use the global RNG in testing at all but use one from the `StableRNGs` package. 

Also, the MLJ `learning_curve` method currently assumes you expose an RNG and not just a seed. It would be a complication to change this, although it's possible other models (eg, models wrapped from other languages, such as XGBoost) will make a change necessary 😄 .

**Comment by jeremiedb** [comment]
Hi @ablaom ,

I just brought changes to support the input of a RNG, following what is being used by DecisionTree.jl. 

In the meantime, I finally managed to have a functionnal GPU support for linear, logistic and Gaussian loss functions, working on both Float32 and Float64. The speedup is not quite what the latest XGBoot gpu_hist brings, but it's still a slight improvement over the cpu approach. There's definitely room for improvement regarding the CPU to GPU communication as well as the kernel efficiency (would have been otherwise suspiciously too easy to beat xgboost gpu!). 

This GPU support however involves a CUDA dependency, which requires Julia 1.4. Considering CUDA is a quite fundamental package, I assume it would be acceptable to change the requirement from Julia 1 to 1.4. It's also my understanding that Julia 1.6 should become the new LTS. 

Pending new version is on gpu-dev branch: 
https://github.com/Evovest/EvoTrees.jl/tree/gpu-dev

**Comment by ablaom** [comment]
> I just brought changes to support the input of a RNG, following what is being used by DecisionTree.jl.

Yay

> In the meantime, I finally managed to have a functionnal GPU support for linear, logistic and Gaussian loss functions, working on both Float32 and Float64. 

Cool! Congratulations.

> I assume it would be acceptable to change the requirement from Julia 1 to 1.4.

Your call. I think MLJFlux has a similar restriction, possible for a similar reason.

> It's also my understanding that Julia 1.6 should become the new LTS.

Can't wait to drop <1.3 support!

I guess you can close this.
